### ISSUE TITLE
Does not work out of box: either "venv: no such file" or "Error: spawn venv ENOENT"

### ISSUE THREAD

**Issue by Zocker1999NET** [matangover/mypy-vscode#70]
Have freshly installed this extension because MS's extension did not detect mypy properly.

But I get strange errors with venv I do not understand not see any settings to tweak around it.

With disabled "Run Using Active Interpreter" (default), following error appears:

```
[2] Check folder: /home/zocker/projects/entertainment-decider/server
[2] Received python path from Python extension: venv
[2] Running dmypy in folder /home/zocker/projects/entertainment-decider/server
/home/zocker/projects/entertainment-decider/server/venv/bin/dmypy --status-file /home/zocker/.config/Code/User/workspaceStorage/5c355c0f430bde2070c52c0e0a14dcaa/matangover.mypy/dmypy-9f254c8cd1b0c78ca11372bd6ed4fc39474e4ac0-148246.json run --log-file /home/zocker/.config/Code/User/workspaceStorage/5c355c0f430bde2070c52c0e0a14dcaa/matangover.mypy/dmypy-9f254c8cd1b0c78ca11372bd6ed4fc39474e4ac0.log -- . --show-column-numbers --no-error-summary --no-pretty --no-color-output --python-executable venv
[2] stdout:
Daemon started

[2] stderr:
mypy: Invalid python executable 'venv': No such file or directory

[2] Error running mypy in /home/zocker/projects/entertainment-decider/server: mypy failed with error: "mypy: Invalid python executable 'venv': No such file or directory". See Output panel for details.
```

With enabled "Run Using Active Interpreter" which is what I want (so the newer mypy 1.0 from the venv is used) following happens:
```
[1] Check folder: /home/zocker/projects/entertainment-decider/server
Activation complete
[1] Received python path from Python extension: venv
[1] Running dmypy in folder /home/zocker/projects/entertainment-decider/server
venv -m mypy.dmypy --status-file /home/zocker/.config/Code/User/workspaceStorage/5c355c0f430bde2070c52c0e0a14dcaa/matangover.mypy/dmypy-9f254c8cd1b0c78ca11372bd6ed4fc39474e4ac0-148246.json run --log-file /home/zocker/.config/Code/User/workspaceStorage/5c355c0f430bde2070c52c0e0a14dcaa/matangover.mypy/dmypy-9f254c8cd1b0c78ca11372bd6ed4fc39474e4ac0.log -- . --show-column-numbers --no-error-summary --no-pretty --no-color-output --python-executable venv
[1] Error running mypy in /home/zocker/projects/entertainment-decider/server: Error: spawn venv ENOENT
```

Both times, VSC was executed from a python venv and it was configured to use it.
The venv was freshly created.

System information:
- Debian GNU/Linux testing/unstable
- Python 3.11.1
- Visual Studio Code 1.75.1

**Comment by matangover** [comment]
If you can give steps how to reproduce the issue that could help. Because on my machine I can't seem to reproduce this. Thanks!

**Comment by Zocker1999NET** [comment]
Oh, by selecting "another" venv I fixed it while making the screenshot you requested. This is the screenshot:
![image](https://user-images.githubusercontent.com/1645646/219865443-cb76c10b-190c-42b8-b3ad-23a247398446.png)

The in blue highlighted variant `Python 3.11.1 ('venv')` was the preselected (and until yesterday probably working) venv I get when launching VSC as follows from shell:
```sh
cd projects/entertainment-decider/server
source venv/bin/activate
vsc  # alias for code .
```

I fixed it by selecting the starred variant `Python 3.11.1 ('venv': venv)`, but I have no idea why. VSC seems to remember the correct venv when restarting it.

**Comment by matangover** [comment]
Thanks for following up. I'll try to see if I can reproduce it.

**Comment by patrys** [comment]
It happens to me constantly now. It seems that on extension start, the extension is getting the `venv` directory from the Python extension instead of Python binary path. Re-selecting the same interpreter fixes the issues as on interpreter updates the correct path is used. Reloading the window or restarting VSCode is enough to make the problem come back.

On extension load:

```
[1] Received python path from Python extension: /path/to/.direnv/python-3.9.16
```

On interpreter change:

```
[1] Received python path from Python extension: /path/to/.direnv/python-3.9.16/bin/python
```

**Comment by matangover** [comment]
Thanks for the report. It seems the Python extension changed their API in a buggy way... I have to look into this

**Comment by matangover** [comment]
@patrys could you please check if the attached version solves it for you? It's using the new API which supposed to be less buggy. I couldn't reproduce the issue myself...

Here it is - install it with the command `Extensions: Install from VSIX`
[mypy-0.3.0.vsix.zip](https://github.com/matangover/mypy-vscode/files/11359650/mypy-0.3.0.vsix.zip)

**Comment by patrys** [comment]
@matangover gave it a try, it seems to work after forcibly reselecting the same interpreter:

```
Mypy extension activated, version 0.3.0
Registering listener for interpreter changed event
Waiting for Python extension to load
Python extension loaded
Listener registered
[1] Check folder: /Users/patrys/GitHub/saleor/saleor
Activation complete
[1] Received Python path from Python extension: /Users/patrys/GitHub/saleor/saleor/.direnv/python-3.9.16/bin/python
[1] Running dmypy in folder /Users/patrys/GitHub/saleor/saleor
/Users/patrys/GitHub/saleor/saleor/.direnv/python-3.9.16/bin/dmypy --status-file '/Users/patrys/Library/Application Support/Code/User/workspaceStorage/061602eef601c65501aa01ce31c7d68d/matangover.mypy/dmypy-e8698e39a789a2864fa24fa537ccf0bac5caac48-21394.json' run --log-file '/Users/patrys/Library/Application Support/Code/User/workspaceStorage/061602eef601c65501aa01ce31c7d68d/matangover.mypy/dmypy-e8698e39a789a2864fa24fa537ccf0bac5caac48.log' -- . --show-column-numbers --no-error-summary --no-pretty --no-color-output --python-executable /Users/patrys/GitHub/saleor/saleor/.direnv/python-3.9.16/bin/python
Active interpreter changed for resource: /Users/patrys/GitHub/saleor/saleor
```

**Comment by patrys** [comment]
Sorry, I was not clear. I had to reselect the interpreter once, after that it seems to work across reloads and restarts.

**Comment by matangover** [comment]
Great. I will release it to everyone

**Comment by underyx** [comment]
I think I'm getting the same error and reselecting the interpreter does not fix it:

```
2023-07-11 09:01:17.355 [info] No interpreter found from setting mypy-type-checker.interpreter
2023-07-11 09:01:17.355 [info] Getting interpreter from ms-python.python extension for workspace /Users/underyx/r2c/app
2023-07-11 09:01:17.357 [info] Interpreter from ms-python.python extension for /Users/underyx/r2c/app: /Users/underyx/r2c/app/server/.venv/bin/python
2023-07-11 09:01:17.358 [info] Using cwd from `python.linting.cwd`.
2023-07-11 09:01:17.361 [info] Server run command: /Users/underyx/r2c/app/server/.venv/bin/python /Users/underyx/.vscode/extensions/ms-python.mypy-type-checker-2023.1.11911749/bundled/tool/lsp_server.py
2023-07-11 09:01:17.361 [info] Server: Start requested.
2023-07-11 09:01:17.449 [info] [Error - 9:01:17 AM] Mypy Type Checker client: couldn't create connection to server.
2023-07-11 09:01:17.449 [info] Launching server using command /Users/underyx/r2c/app/server/.venv/bin/python failed. Error: spawn /Users/underyx/r2c/app/server/.venv/bin/python ENOENT
2023-07-11 09:01:17.449 [error] Server: Start failed: Launching server using command /Users/underyx/r2c/app/server/.venv/bin/python failed. Error: spawn /Users/underyx/r2c/app/server/.venv/bin/python ENOENT
```

even though
```terminal
(semgrep-app-py3.11) underyx@san app % /Users/underyx/r2c/app/server/.venv/bin/python --version
Python 3.11.0
```

**Comment by Alexander-Wilms** [comment]
I encounter the issue as well. However, I accidentally commented on the related issue in the repoository of the ms-python.mypy-type-checker extension:

https://github.com/microsoft/vscode-mypy/issues/85#issuecomment-1846277220

---

I encounter this issue after switching from mypy 1.5.1 installed from the openSUSE repos to version 1.7.1 installed in a venv:

```
alexander@PC:~/Technik/Development/advent_of_code_2023> python3.11 -m venv .
alexander@PC:~/Technik/Development/advent_of_code_2023> ./bin/pip3 install -U mypy
Collecting mypy
  Obtaining dependency information for mypy from https://files.pythonhosted.org/packages/af/9a/ad4b219cf27496653c2312407d6a47593f25f5b53e2d163ade5961cac78c/mypy-1.7.1-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata
  Using cached mypy-1.7.1-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (1.8 kB)
Collecting typing-extensions>=4.1.0 (from mypy)
  Obtaining dependency information for typing-extensions>=4.1.0 from https://files.pythonhosted.org/packages/24/21/7d397a4b7934ff4028987914ac1044d3b7d52712f30e2ac7a2ae5bc86dd0/typing_extensions-4.8.0-py3-none-any.whl.metadata
  Using cached typing_extensions-4.8.0-py3-none-any.whl.metadata (3.0 kB)
Collecting mypy-extensions>=1.0.0 (from mypy)
  Using cached mypy_extensions-1.0.0-py3-none-any.whl (4.7 kB)
Using cached mypy-1.7.1-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (12.4 MB)
Using cached typing_extensions-4.8.0-py3-none-any.whl (31 kB)
Installing collected packages: typing-extensions, mypy-extensions, mypy
Successfully installed mypy-1.7.1 mypy-extensions-1.0.0 typing-extensions-4.8.0

[notice] A new release of pip is available: 23.2.1 -> 23.3.1
[notice] To update, run: python3.11 -m pip install --upgrade pip
alexander@PC:~/Technik/Development/advent_of_code_2023> ./python3.11 -m pip install --upgrade pip
bash: ./python3.11: Datei oder Verzeichnis nicht gefunden
alexander@PC:~/Technik/Development/advent_of_code_2023> ./bin/python3.11 -m pip install --upgrade pip
Requirement already satisfied: pip in ./lib64/python3.11/site-packages (23.2.1)
Collecting pip
  Obtaining dependency information for pip from https://files.pythonhosted.org/packages/47/6a/453160888fab7c6a432a6e25f8afe6256d0d9f2cbd25971021da6491d899/pip-23.3.1-py3-none-any.whl.metadata
  Using cached pip-23.3.1-py3-none-any.whl.metadata (3.5 kB)
Using cached pip-23.3.1-py3-none-any.whl (2.1 MB)
Installing collected packages: pip
  Attempting uninstall: pip
    Found existing installation: pip 23.2.1
    Uninstalling pip-23.2.1:
      Successfully uninstalled pip-23.2.1
Successfully installed pip-23.3.1
alexander@PC:~/Technik/Development/advent_of_code_2023> ./bin/pip3 show mypy
Name: mypy
Version: 1.7.1
Summary: Optional static typing for Python
Home-page: https://www.mypy-lang.org/
Author: Jukka Lehtosalo
Author-email: jukka.lehtosalo@iki.fi
License: MIT
Location: /home/alexander/Technik/Development/advent_of_code_2023/lib64/python3.11/site-packages
Requires: mypy-extensions, typing-extensions
Required-by:
```

settings.json:
```json
[...]
"mypy.dmypyExecutable": "${workspaceFolder}/bin/dmypy",
[...]
```

```
Mypy extension activated, version 0.2.3
Registering listener for interpreter changed event
Listener registered
[1] Check folder: /home/alexander/Technik/Development/advent_of_code_2023
Activation complete
[1] Received python path from Python extension: /usr/bin/python3
[1] Running dmypy in folder /home/alexander/Technik/Development/advent_of_code_2023
/home/alexander/Technik/Development/advent_of_code_2023/bin/dmypy --status-file /home/alexander/.var/app/com.visualstudio.code/config/Code/User/workspaceStorage/6aa9a480d6ce8ba18f940fde3c399e28/matangover.mypy/dmypy-ef52036a7175d483d910011887fe64b438e354e4-169.json run --log-file /home/alexander/.var/app/com.visualstudio.code/config/Code/User/workspaceStorage/6aa9a480d6ce8ba18f940fde3c399e28/matangover.mypy/dmypy-ef52036a7175d483d910011887fe64b438e354e4.log -- . --show-column-numbers --no-error-summary --no-pretty --no-color-output --python-executable /usr/bin/python3
[1] Error running mypy in /home/alexander/Technik/Development/advent_of_code_2023: Error: spawn /home/alexander/Technik/Development/advent_of_code_2023/bin/dmypy ENOENT
```

**Comment by matangover** [comment]
Thanks for reporting. This doesn't seem like the same issue as reported above, because you are specifying the `dmypyExecutable` manually. But it is the same error message. So to debug your issue:
1. Does the file `/home/alexander/Technik/Development/advent_of_code_2023/bin/dmypy` actually exist?
2. If you run it from the Terminal, does it run without error?

**Comment by matangover** [comment]
The original issue reported above is now fixed and released. @Alexander-Wilms if you're still experiencing an issue please open a new issue thanks
